[LLDB][Telemetry]Define TargetInfo for collecting data about a target

oontvoo opened the pull request.

oontvoo committed:
[LLDB][Telemetry]Define TargetInfo for collecting data about a target

oontvoo committed:
Merge branch 'llvm:main' into target

oontvoo committed:
introduce helper

oontvoo commented:
(This PR assumes that we can rely on PR/128534 - that is, we always build the telemetry library. Whether it is enabled is dictated by Config::EnableTelemetry)

oontvoo committed:
undo change

oontvoo committed:
Merge branch 'main' into target

oontvoo committed:
update interface

oontvoo commented:
Hi @labath  @JDevlieghere , just FYI that both this PR and PR/129354 have common change (which is modifying the `ScopedDispatcher`) Feel free to review either of these PR first, but probably easier to not review them at the same time (since you may be repeating the same comment/feedback) .
Thanks!

oontvoo committed:
formatting

oontvoo committed:
remove unused

oontvoo committed:
formatting

oontvoo committed:
qual

oontvoo committed:
use UUID for target_uuid

oontvoo committed:
Merge branch 'llvm:main' into target

oontvoo committed:
Merge branch 'main' into target

oontvoo committed:
revert dup

oontvoo committed:
Merge branch 'target' of github.com:oontvoo/llvm-project into target

oontvoo committed:
Merge branch 'main' into target

oontvoo committed:
unused include

oontvoo committed:
Update Telemetry.h

oontvoo committed:
Update Telemetry.cpp

oontvoo committed:
reconcile with main

oontvoo committed:
Merge branch 'target' of github.com:oontvoo/llvm-project into target

oontvoo committed:
format

labath reviewed (commented):
I find it confusing to use the same entry type for `Target::SetExecutableModule` and `Process::SetExitStatus`. I think it just doesn't match the lldb object model. I suspect you somehow want to bracket the existence of a process between a "start" and "exit" events, but there's no guarantee that this will do that.

SetExecutableModule can be called both before a process is started and after that (multiple times, due to execve). And if you run the same target repeatedly you may get multiple SetExitStatus calls without ever changing the executable module.

I think we need to slow down and figure out what exactly you want to collect, and then find a place to do that.

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Pavel Labath <pavel@labath.sk>

oontvoo commented:
> I find it confusing to use the same entry type for `Target::SetExecutableModule` and `Process::SetExitStatus`. I think it just doesn't match the lldb object model. I suspect you somehow want to bracket the existence of a process between a "start" and "exit" events, but there's no guarantee that this will do that.
> 
> SetExecutableModule can be called both before a process is started and after that (multiple times, due to execve). And if you run the same target repeatedly you may get multiple SetExitStatus calls without ever changing the executable module.
> 
> I think we need to slow down and figure out what exactly you want to collect, and then find a place to do that.

At a high level, we want to know:
What targets were debugged + related startup/exit statistics?
  -  How long did it take to load the target? ( the first time it was load vs subsequent times ...)
  -  Did the target exit successfully? (if not, what was the issue) How long did the "exit" take?

So in that sense we care slightly "less" about a "process" but more about a "target". (We'd still like to keep track of the process as a way to "count' how many times a target (or many targets) was run in a whole debugging session).

Does that make sense?

labath commented:
> Does that make sense?

Somewhat, but I think there's still a lot of confusion between a target vs. process vs. the main executable of the target.

> 
>     * How long did it take to load the target? ( the first time it was load vs subsequent times ...)

Define "target". Define "loading a target". When a process does an execve is it still the same target (as far as the rest of lldb is concerned, it is, even though it starts executing a different binary)?

Right now, you're measuring the time it takes to run SetExecutableModule. I think that's a reasonable thing to measure. What I have a problem with is drawing a line between that and SetExitStatus, as those two things are very different. I also (to a slightly lesser degree) have an issue with calling that "TargetInfo", as it's really information about what we're doing with the executable module (for some target).

In the lldb object model, there's isn't really a concept of "loading a target". Target's can be *created*. They can be created with or without an executable module. If they're created without it, the module can be set at a later time (e.g. during an attach operation). While setting the executable module, the debug info for that module may or may not be parsed, depending on the `preload-symbols` setting. I think this is the definition of "loading" you're referring to, but that is not "loading a target", that is "loading (parsing) the debug info for a module" (which happens to be the executable module of some target). 

If you want to measure SetExecutableModule, then I think you should call it that (SetExecutableModuleInfo?). With that framing, I think everything you have here makes sense. You are reporting that we've set (for whatever reason) the executable module (identified by its UUID) of a target (we don't have a way to identify a target independently of its executable, but it sounds like you don't care about that), and saying it took certain amount of seconds.

>     * Did the target exit successfully? (if not, what was the issue) How long did the "exit" take?

Similarly, I have an issue with this framing, because targets don't exit (processes do). I'd call this ProcessExitInfo, because that's exactly what it's doing: it reports that a process, identified by the UUID of its executable (which, ideally, is one of the executables that we've set earlier, but there isn't an exact 1:1 matching between the two events) has exited and how it did that.

oontvoo commented:
> Define "target". 

"target" == the thing to be debugged. Eg.., `lldb <target>` or (after starting lldb, you'd do `target create <target>`).

> Define "loading a target". When a process does an execve is it still the same target (as far as the rest of lldb is concerned, it is, even though it starts executing a different binary)?

I think by "loading" i mean the time it takes from "target create" command till when lldb command prompt is ready to accept the next command. (Admitted, in this implementation, it's not quite measuring that since that's a bit more tricky to instrument, so I settled to the closest thing, which was around initting the main executable).

Can you clarify your definition of "target" ? :) You're saying a "target" could have *different* main-executable ("different binary")?? How would that use case look? 

I understand you could do something like
```
~/home/ $ lldb  # << start LLDB w/o specifying a debug target
< ... inside lldb prompt >
0: (lldb) target create path/to/MyTarget
1:  (lldb) run  
<.... finish>
2: (lldb) run
< ... finish >
3: (lldb) run
```

Same target, 3 different "runs" but still the same *binary*.
In other words, we would have had 1 target,  3 (different) processes (ie., different PIDs), 3 exit statuses (one for each of those processes).

Is this the right characterisation?
 
So maybe the data can be restructured to:
```
TargetInfo {
  + target_id;
  + target_name; /// (ie., the <path/to/MyTarget> argument)
  + (other related stuff like platform, arch, etc)

  + init_time_duration: Ideally, measuring the time it takes from "target create" till ready. (or just initting the main exec)
  // The following two is less important, but could be interesting
  + start_timestamp: When the target was first created
  + exit_timestamp: when the target is deleted 
}

ProcessInfo {
   + target_id; /// The ID of the Target object that this process belongs to
   + exit_status
}
 
```

labath commented:
> > Define "target".
> 
> "target" == the thing to be debugged. Eg.., `lldb <target>` or (after starting lldb, you'd do `target create <target>`).

Okay. So far, so good.

> 
> > Define "loading a target". When a process does an execve is it still the same target (as far as the rest of lldb is concerned, it is, even though it starts executing a different binary)?
> 
> I think by "loading" i mean the time it takes from "target create" command till when lldb command prompt is ready to accept the next command. (Admitted, in this implementation, it's not quite measuring that since that's a bit more tricky to instrument, so I settled to the closest thing, which was around initting the main executable).

This is where it gets tricky. I know you want that, but I don't think that *all* you want (if it was, you may not even need this patch since it's kind of implicitly measured by the command interpreter telemetry). I think you also want to handle cases where the target is created through SB API.  If *that* was all you wanted, then the best place to instrument might be something like `TargetList::CreateTarget`. However, there is also the case of attaching, in which case you first create a [target without an executable](https://github.com/llvm/llvm-project/blob/665299eb3e7a142199e2c22eb294c5e01ef1655d/lldb/source/Commands/CommandObjectProcess.cpp#L334) and later set the executable [once you've attached to the running process](https://github.com/llvm/llvm-project/blob/665299eb3e7a142199e2c22eb294c5e01ef1655d/lldb/source/Commands/CommandObjectProcess.cpp#L361). There are also other cases (loading a core file (`target create --core`), connecting to a remote debug server (`process connect`)), which look like "attaching" in some ways, but go through slightly different paths.

This is why I suggested way back to focus on SetExecutableModule, as that is (one) place where all of these paths converge, and it covers the most expensive part of the operation. I don't think it's the only way to capture all this information, but I still think it's *a* way to do it. However, it's definitely not the same as "loading/creating a target".

> 
> Can you clarify your definition of "target" ? :) You're saying a "target" could have _different_ main-executable ("different binary")?? How would that use case look?

I can give you at least three:
1. execve
```
$ head *.c
==> ping.c <==
int main() { execlp("./pong", "pong", 0); }

==> pong.c <==
int main() { execlp("./ping", "ping", 0); }
$ lldb ./ping
(lldb) target create "./ping"
Current executable set to '/tmp/ping' (x86_64).
(lldb) process launch --stop-at-entry 
Process 14214 stopped
* thread #1, name = 'ping', stop reason = signal SIGSTOP
    frame #0: 0x00007ffff7fe2bc0 ld-linux-x86-64.so.2`_start
ld-linux-x86-64.so.2`_start:
->  0x7ffff7fe2bc0 <+0>: movq   %rsp, %rdi
    0x7ffff7fe2bc3 <+3>: callq  0x7ffff7fe38c0 ; _dl_start

ld-linux-x86-64.so.2`_dl_start_user:
    0x7ffff7fe2bc8 <+0>: movq   %rax, %r12
    0x7ffff7fe2bcb <+3>: movq   %rsp, %r13
Process 14214 launched: '/tmp/ping' (x86_64)
(lldb) target list
Current targets:
* target #0: /tmp/ping ( arch=x86_64-pc-linux-gnu, platform=host, pid=14214, state=stopped )
### We are executing "ping" now
(lldb) c
Process 14214 resuming
Process 14214 stopped
* thread #1, name = 'pong', stop reason = exec
    frame #0: 0x00007ffff7fe2bc0 ld-linux-x86-64.so.2`_start
ld-linux-x86-64.so.2`_start:
->  0x7ffff7fe2bc0 <+0>: movq   %rsp, %rdi
    0x7ffff7fe2bc3 <+3>: callq  0x7ffff7fe38c0 ; _dl_start

ld-linux-x86-64.so.2`_dl_start_user:
    0x7ffff7fe2bc8 <+0>: movq   %rax, %r12
    0x7ffff7fe2bcb <+3>: movq   %rsp, %r13
(lldb) target list
Current targets:
* target #0: /tmp/pong ( arch=x86_64-pc-linux-gnu, platform=host, pid=14214, state=stopped )
# Same target, same pid, but a different executable -- SetExecutableModule was called to switch the new image
(lldb) c
Process 14214 resuming
Process 14214 stopped
* thread #1, name = 'ping', stop reason = exec
    frame #0: 0x00007ffff7fe2bc0 ld-linux-x86-64.so.2`_start
ld-linux-x86-64.so.2`_start:
->  0x7ffff7fe2bc0 <+0>: movq   %rsp, %rdi
    0x7ffff7fe2bc3 <+3>: callq  0x7ffff7fe38c0 ; _dl_start

ld-linux-x86-64.so.2`_dl_start_user:
    0x7ffff7fe2bc8 <+0>: movq   %rax, %r12
    0x7ffff7fe2bcb <+3>: movq   %rsp, %r13
(lldb) target list
Current targets:
* target #0: /tmp/ping ( arch=x86_64-pc-linux-gnu, platform=host, pid=14214, state=stopped )
# Still the same target, but now we're back to the first binary => Another SetExecutableModule call
```
2. Guessing the wrong binary when attaching:
```
$ lldb /bin/ls
(lldb) target create "/bin/ls"
Current executable set to '/bin/ls' (x86_64).
# One SetExecutableModule call here
(lldb) process attach -p 14329
Process 14329 stopped
* thread #1, name = 'cat', stop reason = signal SIGSTOP
    frame #0: 0x00007f337699c1d1 libc.so.6`read + 17
libc.so.6`read:
->  0x7f337699c1d1 <+17>: cmpq   $-0x1000, %rax ; imm = 0xF000 
    0x7f337699c1d7 <+23>: ja     0x7f337699c230 ; <+112>
    0x7f337699c1d9 <+25>: retq   
    0x7f337699c1da <+26>: nopw   (%rax,%rax)
warning: Executable module changed from "/bin/ls" to "/bin/cat".
# And another one here
```
3. Bare metal debugging where you don't actually have an image to debug (or you just don't know/have access to that image). It's too fiddly to set up now, but in this case you can have a target (and a process), but no executable image (no SetExecutableModule calls)

Neither of these scenarios is particularly common, so I don't think they'll skew the data if you present them as "creating a target". And in a way they're correct, since in e.g., the execve case, you actually have to reparse all of the debug info for the new image so, in a sense, you are creating a new target. However, I think we should be very explicit in the code about what we are doing.

oontvoo committed:
reformatting

oontvoo committed:
Addressed review comments:
 - Use separate entries for Process and Main Executable module
 - Rename for clarity

oontvoo commented:
> This is why I suggested way back to focus on SetExecutableModule, as that is (one) place where all of these paths converge, and it covers the most expensive part of the operation. I don't think it's the only way to capture all this information, but I still think it's _a_ way to do it. However, it's definitely not the same as "loading/creating a target".

Thanks for the example! THat makes a lot more sense now.
I've updated the patch to : 
 - Use separate entries for describing the Main-executable and for the process-exit
 - Rename the struct as suggested

About the `exec*()` use case, would be nice to be able to have *something* to uniquely identify each image ... but i'm not sure what it is ... for now they would just have to be duplicate entries (same module id)

oontvoo committed:
...

oontvoo commented on lldb/source/Target/Process.cpp:
```diff
@@ -1064,6 +1065,29 @@ const char *Process::GetExitDescription() {
+  TargetSP target_sp(Debugger::FindTargetWithProcessID(m_pid));
+  if (target_sp) {
+    helper.SetDebugger(&(target_sp->GetDebugger()));
+    exec_uuid = target_sp->GetExecModuleUUID();
```
We save the Target's exec-module's UUID in the Target object because the Module object is gone by the time we're here.

oontvoo replied:
I can't seem to reproduce it now but I was getting a crash (due to accessing the ModuleSP object here), hence the check on the TargetSP and the code to save it.

But anyway,  i'll revert that back to the simple case. We can fix the crash if/when it manifest again

oontvoo replied:
@labath  Ah, I've found the test that crashed! (See the build log on this PR). Without the check for the validity of the Target pointer and the module pointer, the test crashed/timedout. By contrast, if we added this check, the whole test suite passed:

```
  TargetSP target_sp(Debugger::FindTargetWithProcessID(m_pid));
   if (target_sp) {
    helper.SetDebugger(&(target_sp->GetDebugger()));
    if (ModuleSP mod = target_sp->GetExecutableModule())
      module_uuid = mod->GetUUID();
   }

```

labath replied:
That sort of makes sense. The target would not be available if it was already being destroyed. However, `Debugger::FindTargetWithProcessID` is still a very roundabout way to check for target validity. You should be able to achieve the same thing with:
```
if (TargetSP target_sp = m_target_wp.lock()) {
  ...
}
```

labath commented on lldb/source/Target/Target.cpp:
```diff
@@ -1559,10 +1560,29 @@ void Target::DidExec() {
+      info->exec_mod = executable_sp;
+      info->exec_uuid = m_current_exec_mod_uuid;
+      info->pid = pid;
+      info->arch_name = executable_sp->GetArchitecture().GetArchitectureName();
```
I think this is unnecessarily depriving you of potentially useful information. With just the architecture name, you wouldn't be able to tell a (local?) x86_64 linux debug session from a (remote) x86_64 darwin/freebsd/etc. session. Consider replacing this with `executable_sp->GetArchitecture()->GetTriple().getTriple()`

oontvoo replied:
done

labath reviewed (commented):
This is better, though I'm not particularly fond of the addition of the new Target field. I'd like to see if we could remove that.

oontvoo commented on lldb/source/Target/Process.cpp:
```diff
@@ -1064,6 +1065,29 @@ const char *Process::GetExitDescription() {
+
+  // Find the executable-module's UUID, if available.
+  UUID exec_uuid;
+  // Check if there is (still) a valid target and get the debugger and exec_uuid
+  // from it.
+  TargetSP target_sp(Debugger::FindTargetWithProcessID(m_pid));
+  if (target_sp) {
```
done

oontvoo committed:
addressed review comment

oontvoo committed:
remove unused import

labath reviewed (commented):
I think this is fine now. Time to call in @JDevlieghere :)

oontvoo committed:
Merge branch 'llvm:main' into target

oontvoo commented:
Hi, friendly ping? @JDevlieghere  thanks!

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
   void serialize(llvm::telemetry::Serializer &serializer) const override;
 };
 
+struct ExecModuleInfo : public LLDBBaseTelemetryInfo {
```
```suggestion
struct ExecutableModuleInfo : public LLDBBaseTelemetryInfo {
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
 };
 
+struct ExecModuleInfo : public LLDBBaseTelemetryInfo {
+  lldb::ModuleSP exec_mod;
```
Let's drop the `exec_` prefix as this is relatively obvious from the class it belongs to and we don't consistently prefix all the members. 
```suggestion
  lldb::ModuleSP exec_mod;
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
+  /// The triple of this executable module.
+  std::string triple;
+
+  /// If true, this entry was emitted at the beginning of an event (eg., before
```
```suggestion
  /// If true, this entry was emitted at the beginning of an event (e.g. before
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
+  }
+
+  static bool classof(const TelemetryInfo *T) {
+    // Subclasses of this is also acceptable
```
```suggestion
    // Subclasses of this is also acceptable.
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
+};
+
+struct ProcessExitInfo : public LLDBBaseTelemetryInfo {
+  UUID exec_uuid;
```
```suggestion
  UUID uuid;
```

oontvoo replied:
done

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
+  }
+
+  static bool classof(const TelemetryInfo *T) {
+    // Subclasses of this is also acceptable
```
```suggestion
    // Subclasses of this is also acceptable.
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -146,6 +155,59 @@ struct DebuggerInfo : public LLDBBaseTelemetryInfo {
+
+/// Describes an exit status.
+struct ExitDescription {
+  int exit_code;
```
Contrary to my other comments I think we can keep the `exit_` prefix here as "exit code" is a fairly well understood term.

oontvoo replied:
done

JDevlieghere commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -76,15 +76,36 @@ void CommandInfo::serialize(Serializer &serializer) const {
 }
 
+std::atomic<uint64_t> CommandInfo::g_command_id_seed = 0;
+uint64_t CommandInfo::GetNextId() { return g_command_id_seed.fetch_add(1); }
```
```suggestion
uint64_t CommandInfo::GetNextID() { return g_command_id_seed.fetch_add(1); }
```

oontvoo replied:
done

JDevlieghere commented on lldb/source/Target/Process.cpp:
```diff
@@ -1066,6 +1067,26 @@ const char *Process::GetExitDescription() {
+
+  // Find the executable-module's UUID, if available.
+  Target &target = GetTarget();
+  helper.SetDebugger(&(target.GetDebugger()));
```
```suggestion
  helper.SetDebugger(&target.GetDebugger());
```

oontvoo replied:
done

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -1559,10 +1560,29 @@ void Target::DidExec() {
   ClearModules(false);
 
   if (executable_sp) {
+    lldb::pid_t pid;
+    if (ProcessSP proc = GetProcessSP())
+      pid = proc->GetID();
```
```suggestion
    lldb::pid_t pid = LLDB_INVALID_PROCESS_ID;
    if (ProcessSP proc = GetProcessSP())
      pid = proc->GetID();
```

JDevlieghere reviewed (approved):
LGTM modulo nits.

oontvoo committed:
addressed review comments

oontvoo committed:
Update Telemetry.h

oontvoo committed:
Update Telemetry.cpp

oontvoo committed:
Update CommandInterpreter.cpp

oontvoo committed:
Update Target.cpp

oontvoo committed:
Merge branch 'llvm:main' into target

oontvoo committed:
Merge branch 'target' of github.com:oontvoo/llvm-project into target

oontvoo committed:
more fixing

oontvoo committed:
formating

oontvoo committed:
Merge branch 'main' into target

oontvoo commented:
Thanks!
